src/circuit/gadget/utilities/lookup_range_check.rs

```diff
@@ -0,0 +1,297 @@
+        circuit::{layouter::SingleChipLayouter, Layouter},
+        dev::MockProver,
+        plonk::{Assignment, Circuit, ConstraintSystem, Error},
+    };
+    use pasta_curves::{arithmetic::FieldExt, pallas};
+
+    use std::marker::PhantomData;
+
+    #[test]
+    fn lookup_range_check() {
+        struct MyCircuit<F: FieldExt + PrimeFieldBits> {
+            _marker: PhantomData<F>,
+        }
+
+        impl<F: FieldExt + PrimeFieldBits> UtilitiesInstructions<F> for MyCircuit<F> {
+            type Var = CellValue<F>;
+        }
+
+        impl<F: FieldExt + PrimeFieldBits> Circuit<F> for MyCircuit<F> {
+            type Config = LookupRangeCheckConfig<F, K>;
+
+            fn configure(meta: &mut ConstraintSystem<F>) -> Self::Config {
+                let running_sum = meta.advice_column();
+                let table_idx = meta.fixed_column();
+                let q_lookup = meta.fixed_column();
+
+                let perm = meta.permutation(&[running_sum.into()]);
+
+                LookupRangeCheckConfig::<F, K>::configure(
+                    meta,
+                    q_lookup,
+                    running_sum,
+                    table_idx,
+                    perm,
+                )
+            }
+
+            fn synthesize(
+                &self,
+                cs: &mut impl Assignment<F>,
+                config: Self::Config,
+            ) -> Result<(), Error> {
+                let mut layouter = SingleChipLayouter::new(cs)?;
+
+                // Load table_idx
+                config.load(&mut layouter)?;
+
+                let num_words = 6;
+                let words_and_expected_final_zs = [
+                    (F::from_u64((1 << (num_words * K)) - 1), F::zero()), // a word that is within num_words * K bits long
+                    (F::from_u64(1 << (num_words * K)), F::one()), // a word that is just over num_words * K bits long
+                ];
+
+                for (words, expected_final_z) in words_and_expected_final_zs.iter() {
+                    let expected_zs = expected_zs::<F, K>(*words, num_words);
+
+                    // Load the value to be decomposed into the circuit.
+                    let words = self.load_private(
+                        layouter.namespace(|| "words"),
+                        config.running_sum,
+                        Some(*words),
+                    )?;
+
+                    let zs = layouter.assign_region(
+                        || "word within range",
+                        |mut region| {
+                            for idx in 0..num_words {
+                                // Assign fixed column to activate lookup.
+                                region.assign_fixed(
+                                    || format!("lookup on row {}", idx),
+                                    config.q_lookup,
+                                    idx,
+                                    || Ok(F::one()),
+                                )?;
+                            }
```

Comment:
Why can't this be activated inside `lookup_range_check` (inside the `for (idx, word) in` loop, activated at `offset + idx`)? If it can't be, document that behaviour on `lookup_range_check`.